### PR TITLE
GetAccountDetail pagination tests

### PR DESCRIPTION
**Pull request by MBoldyrev**
Signed-off-by: Mikhail Boldyrev <miboldyrev@gmail.com>

[IR-515](https://jira.hyperledger.org/browse/IR-515)
[POW-091](https://soramitsu.atlassian.net/wiki/spaces/IS/pages/1127317518/POW+N+091+GetAccountDetail+pagination)

### Description of the Change

Tests added for GetAccountDetail pagination.

### Benefits

One step towards pagination of GetAccountDetail query.

### CODE REVIEW

**Comment by igor-egorov** [comment]
👀 
```diff
+425
-0
```

File: test/module/irohad/ametsuchi/postgres_query_executor_test.cpp [hunk 1]
```diff
@@ -992,6 +996,426 @@ namespace iroha {
...
+                                           key.value_or(""),
+                                           writer.value_or(""),
+                                           first_record_writer,
+                                           first_record_key)
+                         .build();
+        return executeQuery(query);
+      }
+
+      /**
+       * Check the page response.
```
Comment (igor-egorov) [inline test/module/irohad/ametsuchi/postgres_query_executor_test.cpp]: Not clear what kind of check it does.
Please add more explanations. It is not obvious due to huge method body.

File: test/module/irohad/ametsuchi/postgres_query_executor_test.cpp [hunk 2]
```diff
@@ -992,6 +996,426 @@ namespace iroha {
...
+
+      void validatePageResponse(
+          const shared_model::interface::AccountDetailResponse &response,
+          boost::optional<std::string> req_writer,
+          boost::optional<std::string> req_key,
+          boost::optional<std::string> first_record_writer,
+          boost::optional<std::string> first_record_key,
+          size_t page_size) {
+        auto optional_match = [](const auto &opt, const auto &val) {
+          return not opt or opt.value() == val;
```
Comment (igor-egorov) [inline test/module/irohad/ametsuchi/postgres_query_executor_test.cpp]: The purpose of the lamda is unclear.
Absence of opt will result in absence of an act of comparison, but the lambda will return true.
Not sure that is the desired behavior.

File: test/module/irohad/ametsuchi/postgres_query_executor_test.cpp [hunk 3]
```diff
@@ -992,6 +996,426 @@ namespace iroha {
...
+        };
+
+        std::map<std::string, std::map<std::string, std::string>> expected_data;
```
Comment (igor-egorov) [inline test/module/irohad/ametsuchi/postgres_query_executor_test.cpp]: I guess it is "writer", "key" and "value". Please a comment.

File: test/module/irohad/ametsuchi/postgres_query_executor_test.cpp [hunk 4]
```diff
@@ -992,6 +996,426 @@ namespace iroha {
...
+        size_t expected_page_size = 0;
+        boost::optional<RecordId> next_record;
+        bool page_started = false;
+        bool page_ended = false;
+        for (const auto &added_writer_and_data : this->added_data_) {
+          const auto &writer = added_writer_and_data.first;
+          const auto &added_data_by_writer = added_writer_and_data.second;
+
+          // check if writer matches query
+          if (optional_match(req_writer, writer)) {
```
Comment (igor-egorov) [inline test/module/irohad/ametsuchi/postgres_query_executor_test.cpp]:
```suggestion
          if (req_writer.value() == writer) {
```

File: test/module/irohad/ametsuchi/postgres_query_executor_test.cpp (hunk 2 above)
Comment (MBoldyrev) [reply test/module/irohad/ametsuchi/postgres_query_executor_test.cpp]: It is needed to filter expected response data. For example, if *writer* was specified in query, we should compare it to match the data row, but if it was not, any row matches.

File: test/module/irohad/ametsuchi/postgres_query_executor_test.cpp (hunk 1 above)
Comment (MBoldyrev) [reply test/module/irohad/ametsuchi/postgres_query_executor_test.cpp]: it is supposed to check everything. anyway, added it to description.

File: test/module/irohad/ametsuchi/postgres_query_executor_test.cpp (hunk 4 above)
Comment (MBoldyrev) [reply test/module/irohad/ametsuchi/postgres_query_executor_test.cpp]: answered above

File: test/module/irohad/ametsuchi/postgres_query_executor_test.cpp [hunk 5]
```diff
@@ -995,6 +999,427 @@ namespace iroha {
...
+      template <typename... Args>
+      auto queryPageAndValidateResponse(Args... args)
+          -> decltype(queryPage(std::declval<Args>()...)) {
+        auto response = queryPage(args...);
+        validatePageResponse(response, args...);
+        return response;
+      }
+    };
+
+    TEST_F(GetAccountDetailPagedExecutorTest, NoPageMetaData) {
```
Comment (igor-egorov) [inline test/module/irohad/ametsuchi/postgres_query_executor_test.cpp]: Please add "given/when/then" or at least short description to the test cases.

File: test/module/irohad/ametsuchi/postgres_query_executor_test.cpp [hunk 6]
```diff
@@ -995,6 +999,427 @@ namespace iroha {
...
+          response, boost::none, boost::none, boost::none, boost::none, 3 * 3);
+    }
+
+    TEST_F(GetAccountDetailPagedExecutorTest, NonExistentFirstRecord) {
+      addDetails(1, 1);
+      checkStatefulError<shared_model::interface::StatefulFailedErrorResponse>(
+          queryPage(boost::none, boost::none, makeAccountId(2), boost::none, 2),
+          kInvalidPagination);
+    }
+
```
Comment (igor-egorov) [inline test/module/irohad/ametsuchi/postgres_query_executor_test.cpp]: Maybe it worth to add some visual divider between different kind of tests?

Something like:
```
   ("`-''-/").___..--''"`-._
   `6_ 6  )   `-.  (     ).`-.__.`)
   (_Y_.)'  ._   )  `._ `. ``-..-`
  _..`--'_..-_/  /--'_.' ,'
(il),-''  (li),'  ((!.-'
```

**Pull request merged**